qrcode_detection_package/main.py

```diff
@@ -1,56 +1,358 @@
+                # if a QR-Code was successfully detected save the image and publish contents on the topic
+
+                # save the image that contains successfully decoded QR-Code
+                timestamp = time.strftime("%Y-%m-%d_%H-%M-%S")
+                img_path = f'images/{timestamp}.jpg'
+                cv2.imwrite(img_path, captured_image)
+                self.get_logger().info(
+                    f"Saved image of detected QR-Code as: {img_path}")
+
+                # create QRCodeInfo message to publish on qr_codes topic
+                msg = QRCodeInfo()
+                msg.time_stamp = self.get_clock().now().to_msg()
+                msg.sender_id = "qr_code_scanner_node"
+                msg.qr_code_content = qr_code_content
+                msg.qrcode_position_x = qrcode_center_x
+                msg.qrcode_position_y = qrcode_center_y
+
+                self.qr_publisher.publish(msg)
+                self.get_logger().info("Published QR code info")
+
+                # create dict for sending job finished message
+                payload = {"marker": str(qr_code_content)}
+                self._job_finished_custom_(CommonNode.EXIT_SUCCESS, payload)
+                self.get_logger().info("Published job finished message")
+
+            else:
+                self.get_logger().info("No QR Code found")
+        else:
+            self.get_logger().info("Could not capture image")
 
 
 def main(args=None):
+    """
+    The main function initializes the node and runs the state machine over the lifetime of the node.
+
+    This function initializes the ROS 2 node, creates an instance of the QRCodeScannerNode class,
+    and starts the image processing loop. Then the state machine decides what the node does depending on the internal state.
+    The states are:
+    - "ready": The node is running and waits for a control message which activates it. There happens no image
+               capturing or qr code scanning in this state
+    - "searching": in this state the node continuously takes images and uses the OpenCV library to scan them for
+                   QR-Codes. If a valid code is found, the nodes publishes its contents and position and switches
+                   back to the state "ready"
+    It handles the cleanup operations before shutting down the node.
+
+    Args: 
+        args: Command-line arguments. Default is None.
+
+    Returns: 
+        None
+    """
     rclpy.init(args=args)
     node_id = 'qr_code_scanner_node'
     qr_code_scanner_node = QRCodeScannerNode(node_id)
 
-    try:
-        qr_code_scanner_node.process_images()
-    finally:
-        qr_code_scanner_node.destroy_node()
-        rclpy.shutdown()
+    # start excecuting state machine until node gets destroyed
+    while True:
+        # check if node is in state "ready"
+        # in this state the node waits for the control message to activate the node
+        if (qr_code_scanner_node.node_state == "ready"):
```

Comment:
Why dont you use Python 3.10s match case here instead of if?

Reply:
thought about it and remembered some kind of debate wether we really can use 3.10 and therefore decided to play it safe with if/elif :D but if its safe that we use 3.10 i will change it

Reply:
Yes, we're using Python 3.10 so you can use this feature (and I do, too)